### PR TITLE
Added docs for ObjectIdDict and WeakKeyDict

### LINKED ISSUE #19498: ObjectIdDict exported but not documented
This type (which I guess is supposed to hash things so that if I insert two references to the same object, that's not allowed?) has no docs at all.

### PR DESCRIPTION
**Pull request by mauro3**
Fixes #19498 
Do I need to build any docs as well?

### CODE REVIEW

**Comment by MichaelHatherly** [comment]
> Do I need to build any docs as well?

Yes, the signatures need to be added somewhere in the `doc/stdlib` files. Might be worth waiting a little until https://github.com/JuliaLang/julia/pull/18588 is merged though.

**Comment by mauro3** [comment]
Thanks, will do.

File: base/associative.jl [hunk 1]
```diff
@@ -275,6 +275,15 @@ push!(t::Associative, p::Pair, q::Pair, r::Pair...) = push!(push!(push!(t, p), q
...
 # hashing objects by identity
 
+"""
+    ObjectIdDict([itr])
+
+`ObjectIdDict()` constructs a hash table where the keys are (always)
+object identities.  Unlike `Dict` it is not parameterized on its key
+and value type and thus its `eltype` is always `Pair{Any,Any}`.
+
+See `Dict` for further help.
```
Comment (mauro3) [inline base/associative.jl]: It's not mentioned how to do that in http://docs.julialang.org/en/latest/manual/documentation/, any pointers?

Comment (kshyatt) [inline base/associative.jl]: Now we are ready for the cross ref!

**Comment by mauro3** [comment]
I made the updates to the stdlib-docs and added the `@ref`.

Inconsistencies I found: 

- `Dict` is listed in `doc/src/stdlib/collections.md` under "General Collections", but it does not implement `endof` (which makes sense as it is not ordered).  Shall we leave it as is or should I file an issue?
- #19534: this means that the docs I provided for `WeakKeyDict` are wrong.  But I suggest to fix that issue instead.  I'll try an put a PR together, maybe wait with merging this until #19534 is resolved.

File: doc/src/stdlib/collections.md [hunk 2]
```diff
@@ -177,17 +178,17 @@ Partially implemented by:
...
+[`WeakKeyDict`](@ref) is a hash table implementation where the keys are weak references to objects, and
 thus may be garbage collected even when referenced in a hash table.
 
 [`Dict`](@ref)s can be created by passing pair objects constructed with `=>()` to a [`Dict`](@ref)
 constructor: `Dict("A"=>1, "B"=>2)`. This call will attempt to infer type information from the
 keys and values (i.e. this example creates a `Dict{String, Int64}`). To explicitly specify types
 use the syntax `Dict{KeyType,ValueType}(...)`. For example, `Dict{String,Int32}("A"=>1, "B"=>2)`.
 
-[`Dict`](@ref)s may also be created with generators. For example, `Dict(i => f(i) for i = 1:10)`.
+[`Dict`](@ref) may also be created with generators. For example, `Dict(i => f(i) for i = 1:10)`.
```
Comment (tkelman) [inline doc/src/stdlib/collections.md]: why removing the plural? Could say "A dict may also be created with a generator" to be singular, but I think it was okay as is

Comment (mauro3) [inline doc/src/stdlib/collections.md]: Sorry, that slipped.  I think, after #19534 it should read: "Associative collections may also..."

**Comment by kshyatt** [comment]
Can this be merged?

**Comment by mauro3** [comment]
It would be good to have a decision on #19566.  If that does not get merged this should be updated slightly.  Let me know what's best.

File: doc/src/stdlib/collections.md (hunk 2 above)
Comment (mauro3) [inline doc/src/stdlib/collections.md]: Addressed now.

**Review by tkelman — APPROVED** [review]
lgtm but I guess should be merged after https://github.com/JuliaLang/julia/pull/19566

**Pull request merged**